File: python/103.Binary Tree Zigzag Level Order Traversal.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

from queue import Queue

class Solution:
    def zigzagLevelOrder(self, root: TreeNode) -> List[List[int]]:
        ret = []
        if root == None:
            return ret 
        q = Queue()
        q.put(root)
        is_reverse = 0
        while(not q.empty()):
            length = q.qsize()
            tmpList = []
            for i in range(length):
                tmpNode = q.get()
                if tmpNode.left != None:
                    q.put(tmpNode.left)
                if tmpNode.right != None:
                    q.put(tmpNode.right)
                tmpList.append(tmpNode.val)
            if is_reverse % 2 == 0:
                ret.append(tmpList)
            else:
                ret.append(tmpList[::-1])
            is_reverse = (is_reverse + 1) %2
        return ret

File: cpp/103. Binary Tree Zigzag Level Order Traversal.cpp
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;


struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>>ret;
        if (root == NULL)
            return ret;
        queue<TreeNode *>_queue;
        _queue.push(root);
        int flag = 0; // left to right
        while(!_queue.empty()){
            int length = _queue.size();
            vector<int>tmp_vec;
            for (int i=0; i<length; i++){
                TreeNode* node = _queue.front();
                _queue.pop();
                tmp_vec.push_back(node->val);
                if (node -> left != NULL)
                    _queue.push(node->left);
                if (node -> right != NULL)
                    _queue.push(node->right);
            }
            if (flag == 1)
                reverse(tmp_vec.begin(), tmp_vec.end());
            ret.push_back(tmp_vec);
            flag = (flag + 1) % 2;
        }
        return ret;
    }
};

